pkg/etcd-backup-operator/controller/s3_backup.go

```diff
@@ -0,0 +1,74 @@
+package controller
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	"github.com/coreos/etcd-operator/pkg/backup"
+	"github.com/coreos/etcd-operator/pkg/backup/backend"
+	"github.com/coreos/etcd-operator/pkg/backup/backupapi"
+	backupS3 "github.com/coreos/etcd-operator/pkg/backup/s3"
+	"github.com/coreos/etcd-operator/pkg/cluster/backupstorage"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+const (
+	tmp = "/tmp"
+)
+
+// TODO: remove this and use backend interface for other options (PV, Azure)
+// handleS3 backups up etcd cluster to s3.
+func (b *Backup) handleS3(clusterName string, s3 *api.S3Source) error {
+	awsDir, so, err := b.setupAWSConfig(s3.AWSSecret)
+	if err != nil {
+		return fmt.Errorf("failed to set up aws config: (%v)", err)
+	}
+	defer os.RemoveAll(awsDir)
+
+	prefix := backupapi.ToS3Prefix(s3.Prefix, b.namespace, clusterName)
+	s3Dir, be, err := makeS3Backend(so, prefix, s3.S3Bucket)
+	if err != nil {
+		return fmt.Errorf("failed to create s3 backend: (%v)", err)
+	}
+	defer os.RemoveAll(s3Dir)
+
+	bm := backup.NewBackupManager(b.kubecli, clusterName, b.namespace, nil, be)
+	// this SaveSnap takes 0 as lastSnapRev to indicate that it
+	// saves any snapshot with revision > 0.
+	_, err = bm.SaveSnap(0)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot (%v)", err)
+	}
+	return nil
+}
+
+func (b *Backup) setupAWSConfig(secret string) (string, *session.Options, error) {
+	awsDir, err := ioutil.TempDir(tmp, "")
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to create aws config/cred dir: (%v)", err)
+	}
+
+	so, err := backupstorage.SetupAWSConfig(b.kubecli, b.namespace, secret, awsDir)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to setup aws config: (%v)", err)
+	}
+
+	return awsDir, so, nil
```

Comment:
awsDir shouldn't be returned?

Reply:
need to return `awsDir` so that it can be removed when handles3 is finished.